dynamiqs/propagator/sepropagator.py

```diff
@@ -0,0 +1,90 @@
+
+    Args:
+        H _(array-like or time-array of shape (...H, n, n))_: Hamiltonian.
+        tsave _(array-like of shape (ntsave,))_: Times at which the propagators
+            are saved. The equation is solved from `tsave[0]` to
+            `tsave[-1]`, or from `t0` to `tsave[-1]` if `t0` is specified in `options`.
+        solver: Solver for the integration. As mentioned above, if this is `None`
+            and the Hamiltonian is constant or piece-wise constant, the propagator is computed
+            by matrix exponential. If it is `None` and the Hamiltonian is not
+            constant or piece-wise constant, the propagator is computed using
+            [`dq.sesolve`][dynamiqs.sesolve] with the solver defaulting to
+            [`dq.solver.Tsit5`][dynamiqs.solver.Tsit5]. If the solver is specified, the propagator
+            is computed using [`dq.sesolve`][dynamiqs.sesolve]. See [`dq.sesolve`][dynamiqs.sesolve]
+            for a list of supported solvers.
+
+        gradient: Algorithm used to compute the gradient.
+        options: Generic options, see [`dq.Options`][dynamiqs.Options].
+
+    Returns:
+        [`dq.PropagatorResult`][dynamiqs.PropagatorResult] object holding the result of the
+            propagator computation. Use the attribute `propagator`
+            to access saved quantities, more details in
+            [`dq.PropagatorResult`][dynamiqs.PropagatorResult].
+    """  # noqa: E501
+    if not options.cartesian_batching:
+        raise ValueError(
+            "Flat batching not supported for propagator. "
+            "Only cartesian batching is supported"
+        )
+    if solver is None and (
+        isinstance(H, ConstantTimeArray) or isinstance(H, PWCTimeArray)
+    ):
+        solver = Expm()
+        solver.assert_supports_gradient(gradient)
+        solver_class = ExpmSolver(tsave, None, H, None, solver, gradient, options)
+        result = solver_class.run()
+        return result
+    else:
+        solver = Tsit5() if solver is None else solver
+        initial_states = eye(H.shape[-1])[..., None]
+        seresult = sesolve(
+            H, initial_states, tsave, solver=solver, gradient=gradient, options=options
+        )
+        if options.save_states:
+            # indices are ...i, t, j. Want to permute them to
+            # t, j, i such that the t index is first and each
+            # column of the propogator corresponds to each initial state
+            ndim = len(seresult.states.shape) - 1
+            perm = list(range(ndim - 3)) + [ndim - 2, ndim - 1, ndim - 3]
+            propagators = jnp.transpose(seresult.states[..., 0], perm)
+        else:
+            # otherwise, sesolve has only saved the final states
+            # so we only need to permute the final two axes
+            propagators = seresult.states[..., 0].swapaxes(-1, -2)
+        saved = Saved(propagators, None, None)
+        return PropagatorResult(tsave, solver, gradient, options, saved, seresult.infos)
```

Comment:
For the sake of consistency with the rest of the codebase, I'd suggest introducing a solver `SEPropagatorSolver` (name TBD) in `dynamiqs/core/` that handles this logic of calling `sesolve` with the appropriate shape.

If we do so, we can have a much simpler structure here, e.g. something like:
```python
    solvers = {
        None: ExpmSolver if isinstance(H, (ConstantTimeArray, PWCTimeArray)) else SEPropagatorTsit5,
        Euler: SEPropagatorEuler,
        Dopri5: SEPropagatorDopri5,
        Dopri8: SEPropagatorDopri8,
        Tsit5: SEPropagatorTsit5,
        Kvaerno3: SEPropagatorKvaerno3,
        Kvaerno5: SEPropagatorKvaerno5,
        Propagator: SEPropagatorPropagator,
    }
    solver_class = get_solver_class(solvers, solver)
    solver.assert_supports_gradient(gradient)
    solver = solver_class(tsave, psi0, H, exp_ops, solver, gradient, options)
    result = solver.run()
    return result
```

Note that this is quite annoying to do today because our structure for the various solvers is not clean. Right now, we have core functions (`sesolve`, `mesolve`, `sepropagator`) that take solvers as input (e.g. `dq.solver.Tsit5`, `dq.solver.Propagator`, ...) and which in turn will use internal solvers (e.g. `SETsit5`, `METsit5`, `SEPropagator`, ...). As you see, we have name collisions and overall unclear structure.

@pierreguilmin I think we should probably rename internal solvers as "integrators" and have better naming conventions, such that the terminology "solver" is only user-facing, while "integrators" can be declined into as many as needed for each combination of core function + solver.

@dkweiss31 I propose keeping this part as is for now, and we'll probably refactor later on.

Reply:
Understood! I had a similar thought that the code structure was a bit inconsistent: in the case of Expm, the actual hard work is done elsewhere in a solver class, otherwise the work is done explicitly in `sepropagator.py`. Happy to refactor once the naming structure is more clear.